ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Cleaner.java

```diff
@@ -96,6 +100,7 @@ public class Cleaner extends MetaStoreCompactorThread {
   static final private String CLASS_NAME = Cleaner.class.getName();
   static final private Logger LOG = LoggerFactory.getLogger(CLASS_NAME);
   private boolean metricsEnabled = false;
+  private Optional<Cache<String, TBase>> metaCache = Optional.empty();
```

Comment:
You could use an abstract boolean useCache() method in MetastoreCompactorThread, (and implement in Initiator and Cleaner respectively) and move the metaCache field and the init logic there. As a result computeIfAbsent should not take metaCache as an argument, and caching could be hidden from the subclasses.

Reply:
I have added 
abstract Optional<Cache<String, TBase>> initializeCache(boolean tableCacheOn); in  MetastoreCompactorThread(Had to made MetastoreCompactorThread abstract class now)

Initiator and Cleaner now implements initializeCache based on tableCacheOn property